### ISSUE TITLE
Classloading issue with testbench

### ISSUE THREAD

**Issue by ErrorProne** [vaadin/quarkus#121]
I've a hard time getting pro features to work in quarkus. It looks like UIUnitTests are not able to load project files.

For a simple cases like this
```java
@QuarkusTest
class ReportViewTest extends UIUnitTest {
    @Test
    void removemelater() {
        final ReportView reportView = navigate(ReportView.class);
    }
}
```

The tests crashes early in Vaadins autodiscover method
```
java.lang.ClassNotFoundException: mytest.vaadinpoc.ReportView
    at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:489)
    at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:466)
    at java.base/java.lang.Class.forName0(Native Method)
    at java.base/java.lang.Class.forName(Class.java:375)
    at com.vaadin.testbench.unit.internal.Routes.autoDiscoverViews(Routes.kt:82)
```

The debugger shows me that the project files are excluded from the debugger which loads the extension (and seems to be used in subsequent calls).

This looks like it can be fixed by changing the Routes.kt:autoDiscoverViews method to this

```kotlin
@JvmOverloads
    fun autoDiscoverViews(vararg packageNames: String? = arrayOf()): Routes = apply {
        val classGraph: ClassGraph = ClassGraph().enableClassInfo()
                .enableAnnotationInfo()
                .acceptPackages(*(packageNames.map { it ?: "" }.toTypedArray()))
        classGraph.scan().use { scanResult: ScanResult ->
            scanResult.getClassesWithAnnotation(Route::class.java.name).mapTo(routes) { info: ClassInfo ->
                ClassLoader.getSystemClassLoader().loadClass(info.name).asSubclass(Component::class.java) <-- !!! breaks here
            }
            scanResult.getClassesImplementing(HasErrorParameter::class.java.name).mapTo(errorRoutes) { info: ClassInfo ->
                ClassLoader.getSystemClassLoader().loadClass(info.name).asSubclass(HasErrorParameter::class.java)
            }
        }

        cleanupErrorRoutes()

        println("Auto-discovered views: $this")
    }
```

So basically replacing the Class.forName calls. I'm however not aware of the side effects this might have. A very wild guess is that quarkus loads the extensions with a separate classloader to isolate the extension and prevent name conflicts. Maybe there is something that can be changed in the extension so that a different classloader is used which is capable of loading project files without touching the code of the testbench.

Environment:
Quarkus 3.0.2
Vaadin 24.+
JDK 17 (corretto)

Links:
Vaadin discord discussion: https://discord.com/channels/732335336448852018/1105486407331479583
Question in quarkus github: https://github.com/quarkusio/quarkus/discussions/33240

**Comment by mcollovati** [comment]
As a workaround, if the views do not need dependency injection, the test can be executed as a "simple" JUnit test, without the usage of the Quarkus test extension.

**Comment by benjaminrau** [comment]
Hi there. Just spent like 6 hours on our team trying to get UI Unit tests running while having CDI available for dependency injection. No chance. 

Tried to workaround the issues Finn pointed out but ending up in the routes not beeing registered.

To me the severity is definitely higher since any view which is not trivial will rely on repositories, eventbus or other services.

I even spent another hour searching for anybody out there who has published an example @QuarkusTest which has CDI + Vaadin UIUnitTest support. Didnt find any.

So to me this is a blocker for any Vaadin implementations on Quarkus.

Anyway thanks for all your afforts.

Cheers Benjamin

**Comment by mcollovati** [comment]
Another potential workaround, but it will probably be not enough as there are a few places where UI Unit testing is using `Class.forName`:

* manually setup Vaadin environment, providing necessary routes
* provide CDI aware `InstantiatorFactory`

```java
@QuarkusTest
class ReportViewTest extends UIUnitTest {

    @Override
    @BeforeEach
    protected void initVaadinEnvironment() {
        Routes routes = new Routes().merge(new Routes(Set.of(ReportView.class), Set.of(), true));
        MockVaadin.setup(routes, MockedUI::new, Set.of(TestInstantiatorFactory.class));
    }

    @Test
    void testMe() {
        final ReportView reportView = navigate(ReportView.class);
        test(reportView.button).click();

        Assertions.assertEquals("ABCD", test(reportView.input).getValue());
        System.out.println(reportView);
    }


    public static class TestInstantiatorFactory implements InstantiatorFactory {
        @Override
        public Instantiator createInstantitor(VaadinService service) {
            BeanManager beanManager = CDI.current().getBeanManager();
            return new QuarkusInstantiator(new DefaultInstantiator(service),
                    beanManager);
        }
    }
}
```

And here the application code under test

```java
@ApplicationScoped
public class MyService {

    public String work() {
        return "ABCD";
    }
}


@Route("xxx")
public class ReportView extends Div {

    final Input input;
    final NativeButton button;

    public ReportView(MyService myService) {
        input = new Input();
        button = new NativeButton("OK", ev -> input.setValue(myService.work()));
        add(input, button);
    }
}
```

I hope this could at least let you set up some basic tests, until a proper integration will be ready

**Comment by ErrorProne** [comment]
> Another potential workaround, but it will probably be not enough as there are a few places where UI Unit testing is using `Class.forName`:
> 
> * manually setup Vaadin environment, providing necessary routes
> * provide CDI aware `InstantiatorFactory`
> 
> ```java
> @QuarkusTest
> class ReportViewTest extends UIUnitTest {
> 
>     @Override
>     @BeforeEach
>     protected void initVaadinEnvironment() {
>         Routes routes = new Routes().merge(new Routes(Set.of(ReportView.class), Set.of(), true));
>         MockVaadin.setup(routes, MockedUI::new, Set.of(TestInstantiatorFactory.class));
>     }
> 
>     @Test
>     void testMe() {
>         final ReportView reportView = navigate(ReportView.class);
>         test(reportView.button).click();
> 
>         Assertions.assertEquals("ABCD", test(reportView.input).getValue());
>         System.out.println(reportView);
>     }
> 
> 
>     public static class TestInstantiatorFactory implements InstantiatorFactory {
>         @Override
>         public Instantiator createInstantitor(VaadinService service) {
>             BeanManager beanManager = CDI.current().getBeanManager();
>             return new QuarkusInstantiator(new DefaultInstantiator(service),
>                     beanManager);
>         }
>     }
> }
> ```
> 
> And here the application code under test
> 
> ```java
> @ApplicationScoped
> public class MyService {
> 
>     public String work() {
>         return "ABCD";
>     }
> }
> 
> 
> @Route("xxx")
> public class ReportView extends Div {
> 
>     final Input input;
>     final NativeButton button;
> 
>     public ReportView(MyService myService) {
>         input = new Input();
>         button = new NativeButton("OK", ev -> input.setValue(myService.work()));
>         add(input, button);
>     }
> }
> ```
> 
> I hope this could at least let you set up some basic tests, until a proper integration will be ready

Interesting. I'll try it out ASAP and report back

**Comment by benjaminrau** [comment]
Thanks for your help. Unfortunately `BaseUIUnitTest.getCurrentView` throws a IndexOutOfBoundsExceptions for `verifyAndGetUI().getInternals().getActiveRouterTargetsChain().get(0)`.

Ive taken a similar approach few days ago and experience the same exception.

Also tried to get some inspiration from the following implementation - but it lead me to a solution yet:
https://github.com/mvysny/vaadin-quarkus/blob/main/src/test/java/org/acme/servlet/MockServlet.java

**Comment by mcollovati** [comment]
@benjaminrau in my example, I have no issues calling `getCurrentView()` after `navigate(...)`
Can you provide a sample code that reproduces your issue?

**Comment by ErrorProne** [comment]
@mcollovati please correct me if I'm wrong. Besides the spring integration there is no actual implementation for auth in the mockservlets? At least intellijs usage search tells me that the principal in requests is set only in the SpringServlet.

My first try was to use the same mechanic and implement a MockRequestCustomizer, but this failed. It looks like this
```kotlin
service.context.getAttribute(Lookup::class.java)
                .lookup(MockRequestCustomizer::class.java)?.apply(mockRequest)
```

in the MockVaadin.kt only looks for customizers in the vaadin package, which is very sad since this would have been an awesome way to do things in tests.

In my second try I implemented a very basic MockQuarkusVaadinServlet which uses reflection to set the principal on the request. Which looks like it works but feels so damn hacky. I'll now try to integrate the quarkus @TestSecurity annotation and post the result here later.

**Comment by mcollovati** [comment]
@ErrorProne you're right. Currently security support has (and `MockRequestCustomizer`) has been implemented only for Spring.
It would be worth it to create an issue on Testbench repository to request that feature.

Below you a find a workaround to plug in the `MockRequestCustomizer` to inject security stuff

```java

@QuarkusTest
class MyTest extends UIUnitTest {

    @Override
    @BeforeEach
    protected void initVaadinEnvironment() {
        Routes routes = new Routes().merge(new Routes(Set.of(ReportView.class), Set.of(), true));
        MockVaadin.setup(routes, MockedUI::new, Set.of(TestLookupInitializer.class, TestInstantiatorFactory.class));
    }

    public static class TestInstantiatorFactory implements InstantiatorFactory {

        @Override
        public Instantiator createInstantitor(VaadinService service) {
            BeanManager beanManager = CDI.current().getBeanManager();
            return new QuarkusInstantiator(new DefaultInstantiator(service),
                    beanManager);
        }
    }

    public static class TestLookupInitializer extends LookupInitializer {
        @Override
        public void initialize(VaadinContext context, Map<Class<?>, Collection<Class<?>>> services, VaadinApplicationInitializationBootstrap bootstrap) throws ServletException {
            ensureService(services, MockRequestCustomizer.class, QuarkusSecurityCustomizer.class);
            super.initialize(context, services, bootstrap);
        }
    }

    public class QuarkusSecurityCustomizer implements MockRequestCustomizer {

        @Override
        public void apply(@NotNull MockRequest mockRequest) {
            // security customizations, such as
            // mockRequest.setUserPrincipalInt(...);
            // mockRequest.setUserInRole(...);
        }
    }

}
```

**Comment by ErrorProne** [comment]
Thanks, with my tinkering I've implemented it this way:

```java
MockVaadin.INSTANCE.setMockRequestFactory((MockHttpSession session) -> {
			MockRequest mockRequest = new MockRequest(session);
			mockRequest.setUserPrincipalInt(new PrincipalWithRoles(securityIdentity));
			mockRequest.setUserInRole(PutawayOrderListViewTest::isGranted);
			return mockRequest;
		});
```

Is there a downside to this approach or is the one with the LookupInitializer just the preferred way?

Also there has to be some quarkus specific stuff since QuarkusPrincipal has no roles like the one from spring does. But the method in the UserInRole takes only the Principal and target role.

```java
	private static class PrincipalWithRoles extends QuarkusPrincipal {
		@Getter
		Set<String> roles;

		public PrincipalWithRoles(SecurityIdentity identity) {
			super(identity.getPrincipal().getName());
			this.roles = identity.getRoles();
		}
	}

	/**
	 * Very simple check to test if a user has a certain role
	 */
	private static boolean isGranted(Principal principal, String role) {
		if (!(principal instanceof PrincipalWithRoles)) {
			return false;
		}

		return ((PrincipalWithRoles)principal).getRoles().contains(role);
	}
```

Should this part of the testbench implementation? I would create a ticket in that repository than.

**Comment by mcollovati** [comment]
Your solution seems safe to me.
Using `MockRequestCustomizer` would probably be a cleaner approach, since it allows hooking into the request without replacing the core components, but unfortunately, it works out-of-the-box only for spring at the moment.

For the roles check, maybe the `setUserInRole` can be implemented based on `SecurityIdentity::hasRole` (not tested, just an idea)

```java
mockRequest.setUserPrincipalInt(sec.getPrincipal());
mockRequest.setUserInRole( (p,r) -> {
    return !sec.isAnonymous() && sec.getPrincipal().equals(p) && sec.hasRole(r);
    // or maybe just sec.hasRole(r)?
});
```


Having support for Quarkus in testing would be a nice feature for Testbench UI Unit Test, so creating a ticket may be worth it.

**Comment by ErrorProne** [comment]
Sonuds like an idea. I'll look into it. Right now it is a very hacky implementation anyway and we have to find a way to combine the UiUnitTest-Class and our normal QuarkusIsolatedTest class and clean everything up. But that should be the easier part.

> Having support for Quarkus in testing would be a nice feature for Testbench UI Unit Test, so creating a ticket may be worth it.

Most definitly, I'm currently writing one. When we switched to Vaadin for our current project we pro-actively talked to some Vaadin representatives very explicitly stating that we intend to du quarkus and do a lot of testing. Which has been told to us is no problem. To say our Vaadin experience got quite a bump when we realized that the pro features we pay and asked for do not work out of the box, or are even document in that way, would be an understatement.

@mcollovati thank you very much for your support and hints, they where very helpful and put us into the right direction.
